Add push to docker registry github workflow

pgressa opened the pull request:
Examples of generated projects:
- https://github.com/pgressa/gradle-starter-test
- https://github.com/pgressa/maven-starter-test
- https://github.com/pgressa/maven-graalvm-starter-test
- https://github.com/pgressa/gradle-graalvm-starter-test

graemerocher commented:
Could you fix the conflicts?

alvarosanchez commented on starter-core/src/main/java/io/micronaut/starter/feature/github/workflows/docker/templates/gradleGraalCEWorkflow.rocker.raw:
```diff
@@ -0,0 +1,46 @@
+        uses: DeLaGuardo/setup-graalvm@@2.0
+        with:
+        @if (jdkVersion == JdkVersion.JDK_8) {
+            graalvm-version: 20.2.0.java8
```
20.3 is coming next week, and we would need to update this versions here.

I think you can use `VersionInfo` from the feature and pass it as an argument to the template.

alvarosanchez commented on starter-core/src/main/java/io/micronaut/starter/feature/github/workflows/docker/templates/mavenJavaWorkflow.rocker.raw:
```diff
@@ -0,0 +1,34 @@
+      - name: Build and push docker image
+        env:
+          DOCKER_REGISTRY: ${{ secrets.DOCKER_REGISTRY }}
+        run: ./mvnw clean package jib:build
```
No need to clean, or @aalmiray will appear here 😅

pgressa replied:
oh sorry for this one, copy-paster left over

alvarosanchez commented on starter-core/src/test/groovy/io/micronaut/starter/feature/github/workflows/docker/DockerRegistryWorkflowSpec.groovy:
```diff
@@ -0,0 +1,74 @@
+        when:
+        def output = generate([DockerRegistryWorkflow.NAME])
+        def readme = output['README.md']
+        println(readme)
```
Please remove

alvarosanchez commented on starter-core/src/test/groovy/io/micronaut/starter/feature/github/workflows/docker/GraalVMDockerRegistryWorkflowSpec.groovy:
```diff
@@ -0,0 +1,76 @@
+        when:
+        def output = generate([GraalVMDockerRegistryWorkflow.NAME])
+        def readme = output['README.md']
+        println(readme)
```
Please remove

alvarosanchez commented on starter-core/src/main/java/io/micronaut/starter/feature/build/maven/templates/pom.rocker.raw:
```diff
@@ -1375,7 +1375,18 @@ List<Property> properties
+          </to>
+        </configuration>
+      </plugin>
+}else if (features.contains("jib")) {
```
Right, I see now there are conflicts. Please fix them

pgressa commented on starter-core/src/main/java/io/micronaut/starter/feature/github/workflows/docker/templates/mavenGraalCEWorkflow.rocker.raw:
```diff
@@ -0,0 +1,42 @@
+        run: |
+          ./mvnw package -Dpackaging=docker-native
+          PROJECT_VERSION=`mvn help:evaluate -Dexpression=project.version -q -DforceStdout`
+          IMAGE="${DOCKER_REGISTRY}/@project.getName():${PROJECT_VERSION}"
```
Trying here to stay compatible with gradle version, where docker image with version tag is pushed, meaning latest is not pushed at all.

pgressa replied:
Also once https://github.com/micronaut-projects/micronaut-maven-plugin/issues/48 and https://github.com/micronaut-projects/micronaut-maven-plugin/issues/49 are fixed, the bash part can be removed.

pgressa commented on starter-core/src/main/java/io/micronaut/starter/feature/graalvm/GraalVMFeatureValidator.java:
```diff
@@ -35,7 +36,7 @@ public void validatePreProcessing(Options options, ApplicationType applicationTy
     @Override
     public void validatePostProcessing(Options options, ApplicationType applicationType, Set<Feature> features) {
-        if (features.stream().anyMatch(f -> f instanceof GraalVM)) {
+        if (features.stream().anyMatch(f -> f instanceof GraalVM || f instanceof GraalVMDockerRegistryWorkflow)) {
```
I do not rely on graalvm feature now in any of the workflow features, but for graal the validator is reused.

pgressa commented on starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/templates/gradlePlugins.rocker.raw:
```diff
@@ -13,7 +13,7 @@ Features features
 plugins {
 @if (features.contains("asciidoctor")) {
-    id("org.asciidoctor.jvm.convert") version "3.2.0"
+    id("org.asciidoctor.jvm.convert") version "3.3.0"
```
Should this upgrade go in 2.2. or in separated PR?

alvarosanchez replied:
It is fine here IMO

pgressa commented:
Waiting now for Alvaro to finish support of https://github.com/micronaut-projects/micronaut-maven-plugin/issues/48 and I'm working now on integration test with github

graemerocher commented on starter-core/src/main/java/io/micronaut/starter/feature/github/workflows/docker/GraalVMDockerRegistryWorkflow.java:
```diff
@@ -0,0 +1,83 @@
+ * @since 2.2
+ */
+@Singleton
+public class GraalVMDockerRegistryWorkflow extends AbstractDockerRegistryWorkflow {
```
Think this should be a hidden feature that is automatically applied by default.

graemerocher commented on starter-core/src/main/java/io/micronaut/starter/feature/github/workflows/docker/AbstractDockerRegistryWorkflow.java:
```diff
@@ -0,0 +1,41 @@
+        return Arrays.asList(
+                new Secret("DOCKER_USERNAME", "Docker registry username."),
+                new Secret("DOCKER_PASSWORD", "Docker registry password."),
+                new Secret("DOCKER_REGISTRY", "Docker image registry, for image `foo/bar:latest` it is `foo`."),
```
This should be `DOCKER_ORGANIZATION` since on each cloud this is different (GCP it is the project id, OCI it is the tenancy id + repository name, on Github this is the `OWNER`)

pgressa commented on starter-core/src/main/java/io/micronaut/starter/feature/github/workflows/Secret.java:
```diff
@@ -0,0 +1,51 @@
+ * @author Pavol Gressa
+ * @since 2.2
+ */
+public class Secret {
```
The purpose of having this object here is to automatically create the secrets using starter api

pgressa commented on starter-core/src/main/java/io/micronaut/starter/feature/github/workflows/docker/templates/gradleGraalCEWorkflow.rocker.raw:
```diff
@@ -0,0 +1,42 @@
+          DOCKER_ORGANIZATION: ${{ secrets.DOCKER_ORGANIZATION }}
+          DOCKER_REGISTRY_URL: ${{ secrets.DOCKER_REGISTRY_URL }}
+          TESTCONTAINERS_RYUK_DISABLED: true
+        export DOCKER_IMAGE=`echo "${DOCKER_REGISTRY_URL}/${DOCKER_ORGANIZATION}/@project.getName()" | sed -e 's#//#/#' -e 's#^/##'`
```
By default we want to support ideally all possible docker registries. For this purpose two secrets (`DOCKER_REGISTRY_URL` and `DOCKER_ORGANIZATION`) are created from which the final DOCKER_IMAGE name is composed, plust the DOCKER_REGISTRY_URL is used for login into docker registry.

One could argue that simple concatenation like `$DOCKER_REGISTRY_URL/$DOCKER_ORGANIZATION/$project.name` inside pom.xml or build.gradle would be enough. Unfortunately the opposite is the truth because such chaining would cause creation of invalid docker image name:
- dockerhub - no need to specify uri, so the final image is `/pgressa/foo`
- amazon ecr - in case of flat docker registry hierarchy , the DOCKER_ORGANIZATION is not set, so the final image is `<aws-account-number>.dkr.ecr.<region>.amazonaws.com//foo:0.1`

B/c of the above the final image name is easier to be done in bash than in gradle/pom.

graemerocher replied:
Does this actually work? looks like invalid YAML

pgressa replied:
true.. missed this one, the final test suite is still running

pgressa replied:
for review purpose I'll generate all 4 combinations and reference it from description so you can check

pgressa commented on test-utils/src/main/groovy/io/micronaut/starter/test/github/WorkflowSpec.groovy:
```diff
@@ -0,0 +1,171 @@
+        return beanContext.getBean(GitHubApiClient)
+    }
+
+    protected void pushToGitHub(Project project, List<Secret> secrets){
```
This basically duplicates in a lot the starter-api code of push-to-github. Would like to hear your opinion what is the preferred way:
A) dependency of `starter-api` in `test-utils`
B) moving the shared code into `starter-core`
C) create new module `starter-github`

cc: @graemerocher @alvarosanchez ?

graemerocher replied:
My opinion would be B)

pgressa replied:
ok will move the code today .. please check the rest of the code

pgressa commented on test-core/src/test/groovy/io/micronaut/starter/core/test/feature/github/workflow/DockerRegistryWorkflowSpec.groovy:
```diff
@@ -0,0 +1,113 @@
+ *     <li>DOCKER_ORGANIZATION - Docker organisation repository path.</li>
+ * </ul>
+ */
+@Requires({System.getenv().containsKey(GH_TOKEN) && System.getenv().containsKey(DOCKER_USERNAME) \
```
It would be great to configure these secrets in GitHub project in order to run this test at least against dockerhub:
- the repositories have to exists in docker hub
- ! ideally for GH_TOKEN to create some system account b/c the token needs to have permission to delete projects what is quite dangerous operation.

alvarosanchez replied:
We can have a different token than `GH_TOKEN`. Can you create an issue so that we don't forget? I think it's worth to test it, but I would not hold this PR because of that.

alvarosanchez commented on starter-core/src/main/java/io/micronaut/starter/feature/github/workflows/docker/templates/mavenJavaWorkflow.rocker.raw:
```diff
@@ -0,0 +1,39 @@
+          DOCKER_REGISTRY_URL: ${{ secrets.DOCKER_REGISTRY_URL }}
+        run: |
+          DOCKER_IMAGE=`echo "${DOCKER_REGISTRY_URL}/${DOCKER_ORGANIZATION}/@project.getName()" | sed -e 's#//#/#' -e 's#^/##'`
+          ./mvnw deploy -Dpackaging=docker -Dpackaging=docker-native -Djib.docker.image=$DOCKER_IMAGE
```
Packaging is twice

pgressa replied:
damn.. typo sry

pgressa commented:
@alvarosanchez @graemerocher guys, hopefully addressed all issues. Run the gradle maven combination with graalvm and regular java against dockerhub, OCIR and GCLOUD. There are some tests failing but I would say it's regression from elsewhere, maybe https://github.com/micronaut-projects/micronaut-core/pull/4545

pgressa commented on starter-api/build.gradle:
```diff
@@ -12,7 +12,6 @@ dependencies {
     api "io.micronaut:micronaut-http"
     api "io.micronaut:micronaut-http-client"
     api "io.micronaut:micronaut-validation"
-    api 'org.eclipse.jgit:org.eclipse.jgit:5.9.0.202009080501-r'
```
moved to `starter-core`
